### DISCUSSION TITLE
What is the Best Practices for Implementing Power Switch Control in Custom Zephyr Boards

### DISCUSSION THREAD

**Discussion by zafersn** [zephyrproject-rtos/zephyr#83183] [Q&A]
I have a custom board with a power switch that controls a couple of sensors/devices. I'm looking for guidance on the best practices for integrating this power switch without modifying the main Zephyr repository's YAML and source files.

Here is the schematic for the power switches:
![image](https://github.com/user-attachments/assets/538f31a0-1ba9-4b5e-bb80-5ca8a577ba32)

1. DTS Design: What is the recommended DTS design for incorporating the power switch? I want to avoid editing the sensor's YAML file and driver source files, as this approach could lead to extensive changes across various projects and board configurations. Additionally, it would affect the default official Zephyr release, impacting all teams and users.

2. Driver Implementation: How should I implement the power switch control in the sensor's driver file? Do I need to edit driver files for this support? I know that many sensor drivers use the PM_DEVICE_DT_INST_DEFINE() macro, which allows applications to manage power states by calling pm_device_action_run(device, PM_DEVICE_ACTION_SUSPEND) and pm_device_action_run(device, PM_DEVICE_ACTION_RESUME). 
One idea that came into my mind is to create a general power controller source file with a higher initialization priority than any CONFIG_SENSOR_INIT_PRIORITY. This would allow the power controller to turn on the sensors before their initialization begins. However, this approach might not be optimal without directly associating the power switch with the driver instance.

 Any recommendations on how to proceed with this implementation would be greatly appreciated.

**Comment by unsanded** [comment]
I have not found any mechanism for this (though i am by no means a zephyr veteran). What we do is add `zephyr-deferred-init` to the any device that is power-switched. Then we have to do device_init every time we switch it on. 
This is far from ideal and would definitely be interested in building (or finding out about 😉 ) a tool that inits each device when it is switched on.
